nep-Y.mediawiki

```diff
@@ -0,0 +1,98 @@
+
+==Motivation==
+
+While the original [[nep-11.mediawiki|NEP-11]] already defines this method, it
+lacks some details relevant specifically to the receiver contract and it doesn't
+allow for easy reference. This standard can be used in <code>supportedstandards</code>
+section of [[nep-15.mediawiki|NEP-15 manifest]] or in any other situation to
+refer to contracts able to receive [[nep-11.mediawiki|NEP-11]] tokens (but not
+necessarily being [[nep-11.mediawiki|NEP-11]] themselves). Typical use cases
+for this functionality are: depositing/staking (with contract locking and
+managing tokens according to some rules) or exchange/minting (contract accepting
+transfer to itself, but returning some different token to the sender).
+
+==Specification==
+
+Smart contracts that need to receive and own [[nep-11.mediawiki|NEP-11]] tokens
+MUST implement <code>onNEP11Payment</code> method with the following signature:
+
+<pre>
+{
+  "name": "onNEP11Payment",
+  "parameters": [
+    {
+      "name": "from",
+      "type": "Hash160"
+    },
+    {
+      "name": "amount",
+      "type": "Integer"
+    },
+    {
+      "name": "tokenId",
+      "type": "ByteString"
+    },
+    {
+      "name": "data",
+      "type": "Any"
+    }
+  ],
+  "returntype": "Void"
+}
+</pre>
+
+A contract that doesn't implement this method will not be able to receive
+[[nep-11.mediawiki|NEP-11]] tokens (transfers to its address will fail). This
+method is called by [[nep-11.mediawiki|NEP-11]] contracts which tokens are being
+transferred. A particular token can be obtained via the <code>System.Runtime.GetCallingScriptHash</code>
+system call. Contracts SHOULD check for this hash to match their expectations
+like allowed/accepted tokens, processing this call means that contract trusts
+the caller to implement [[nep-11.mediawiki|NEP-11]] correctly.
+
+<code>from</code> parameter specifies the sender of tokens and can be <code>null</code>
+for minted tokens.
+
+<code>amount</amount> is the amount of tokens transferred to the contract. It's
+always 1 for non-divisible [[nep-11.mediawiki|NEP-11]] tokens.
+
+<code>tokenId</amount> is the transferred token ID.
+
+<code>data</data> is the same <code>data</code>parameter that was given to the
+[[nep-11.mediawiki|NEP-11]] <code>transfer</code> call. Contract can interpret
+it in any application-specific way including imposing any restrictions on
+accepted <code>data</data> contents.
+
+If the contract implementing this NEP can't accept the transfer (for reasons like
+unexpected token, wrong amount, forbidden sender, wrong <code>data</code>
+parameter or any other) it MUST call <code>ABORT</code> or <code>ABORTMSG</code>
+VM opcode to fail the transaction. This is the only way to ensure contract
+won't own transferred tokens, because the method doesn't return any value
+and exceptions can be handled by calling contract.
+
+==Rationale==
+
+This NEP just follows and explains [[nep-11.mediawiki|NEP-11]] semantics, no
+design decisions made here.
```

Comment:
OK, I see that we're not going to change the code. But maybe write a couple of words about the behaviour mentioned in the previous comment?

Reply:
Isn't it already there? SupportedStandards are purely informational until we at least do https://github.com/neo-project/neo/issues/1883.